flink-tests/src/test/java/org/apache/flink/test/checkpointing/utils/StatefulJobSavepointMigrationITCase.java

```diff
@@ -0,0 +1,658 @@
+
+		public static final String SUCCESSFUL_RESTORE_CHECK_ACCUMULATOR = CheckingRestoringNonParallelSourceWithListState.class + "_RESTORE_CHECK";
+
+		private volatile boolean isRunning = true;
+
+		private final int numElements;
+
+		public CheckingRestoringNonParallelSourceWithListState(int numElements) {
+			this.numElements = numElements;
+		}
+
+		@Override
+		public void snapshotState(FunctionSnapshotContext context) throws Exception {
+
+		}
+
+		@Override
+		public void initializeState(FunctionInitializationContext context) throws Exception {
+			ListState<String> unionListState = context.getOperatorStateStore().getListState(
+				CheckpointedNonParallelSourceWithListState.stateDescriptor);
+
+			if (context.isRestored()) {
+				assertThat(unionListState.get(),
+					containsInAnyOrder(
+						CheckpointedNonParallelSourceWithListState.checkpointedString,
+						CheckpointedNonParallelSourceWithListState.checkpointedString1,
+						CheckpointedNonParallelSourceWithListState.checkpointedString2,
+						CheckpointedNonParallelSourceWithListState.checkpointedString3));
+
+				getRuntimeContext().addAccumulator(SUCCESSFUL_RESTORE_CHECK_ACCUMULATOR, new IntCounter());
+				getRuntimeContext().getAccumulator(SUCCESSFUL_RESTORE_CHECK_ACCUMULATOR).add(1);
+			} else {
+				throw new RuntimeException(
+					"This source should always be restored because it's only used when restoring from a savepoint.");
+			}
+		}
+
+		@Override
+		public void run(SourceContext<Tuple2<Long, Long>> ctx) throws Exception {
+
+			// immediately trigger any set timers
+			ctx.emitWatermark(new Watermark(1000));
+
+			synchronized (ctx.getCheckpointLock()) {
+				for (long i = 0; i < numElements; i++) {
+					ctx.collect(new Tuple2<>(i, i));
+				}
+			}
+
+			while (isRunning) {
+				Thread.sleep(20);
+			}
+		}
+
+		@Override
+		public void cancel() {
+			isRunning = false;
+		}
+	}
+
+	private static class CheckpointedParallelSourceWithUnionListState
+		extends RichSourceFunction<Tuple2<Long, Long>> implements CheckpointedFunction {
+
+		final static ListStateDescriptor<String> stateDescriptor =
+			new ListStateDescriptor<>("source-state", StringSerializer.INSTANCE);
+
+		final static String[] checkpointedStrings = {
+			"Here be dragons!",
+			"Here be more dragons!",
+			"Here be yet more dragons!",
+			"Here be the mostest dragons!" };
+
+		private static final long serialVersionUID = 1L;
+
+		private volatile boolean isRunning = true;
+
+		private final int numElements;
+
+		private transient ListState<String> unionListState;
+
+		public CheckpointedParallelSourceWithUnionListState(int numElements) {
+			this.numElements = numElements;
+		}
+
+		@Override
+		public void snapshotState(FunctionSnapshotContext context) throws Exception {
+			unionListState.clear();
+
+			for (String s : checkpointedStrings) {
+				if (s.hashCode() % getRuntimeContext().getNumberOfParallelSubtasks() == getRuntimeContext().getIndexOfThisSubtask()) {
```

Comment:
could replace the conditional block with `unionListState.add(checkpointedStrings[getRuntimeContext().getIndexOfThisSubtask()]`.

This would make the behavior of each subtask obvious from the code.